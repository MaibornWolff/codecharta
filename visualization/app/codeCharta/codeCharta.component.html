<tool-bar-component></tool-bar-component>

<file-extension-bar-component></file-extension-bar-component>
<div class="ribbon-bar-wrapper">
	<ribbon-bar-component class="ribbon-bar-component"></ribbon-bar-component>
</div>

<code-map-component></code-map-component>

<legend-panel-component></legend-panel-component>

<loading-gif-file-component></loading-gif-file-component>

<div id="mw-logo">
	<div class="logo">
		<a href="https://www.maibornwolff.de" target="_blank"><img src="assets/logo.png" alt="MaibornWolff"/></a>
		<h2>
			<span>CodeCharta {{ $ctrl._viewModel.version }}</span>
		</h2>
	</div>
</div>

<node-context-menu-component></node-context-menu-component>
