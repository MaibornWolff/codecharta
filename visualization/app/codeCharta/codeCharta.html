<div id="hud" class="container">

    <div class="row">

        <div class="col l3 m6 s12 col-backdrop">
            <legend-panel-component></legend-panel-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <revision-chooser-component></revision-chooser-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <file-chooser-panel-directive></file-chooser-panel-directive>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <settings-panel-component></settings-panel-component>
        </div>

    </div>

    <div class="row bottom-row">

        <div class="col l3 m6 s12 col-backdrop">
            <detail-panel-component></detail-panel-component>
        </div>

    </div>

    <div id="titlebox">
        <a class="mwlink" href="https://www.maibornwolff.de" target="_blank"><img src="./logo.png" alt="MaibornWolff"></a>
        <h1 id="title"><span>CodeCharta {{$ctrl.pkg.version}}</span></h1>
    </div>

</div>

<code-map-component></code-map-component>