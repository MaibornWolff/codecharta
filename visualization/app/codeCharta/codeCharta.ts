"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent, CodeChartaController} from "./codeCharta.component";

import angular from "angular";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
);


