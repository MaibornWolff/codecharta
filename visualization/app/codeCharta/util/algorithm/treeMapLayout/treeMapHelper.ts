import { getMapResolutionScaleFactor, getMarkingColor, isLeaf } from "../../codeMapHelper"
import { CcState, CodeMapNode, ColorMode, Node } from "../../../codeCharta.model"
import { Vector3 } from "three"
import { CodeMapBuilding } from "../../../ui/codeMap/rendering/codeMapBuilding"
import { HierarchyRectangularNode } from "d3-hierarchy"
import { searchedNodePathsSelector } from "../../../state/selectors/searchedNodes/searchedNodePaths.selector"
import { gradientCalculator } from "../../color/gradientCalculator"
import {
    MetricMinMax,
    selectedColorMetricDataSelector
} from "../../../state/selectors/accumulatedData/metricData/selectedColorMetricData.selector"

export const treeMapSize = 250

const FOLDER_HEIGHT = 2
const MIN_BUILDING_HEIGHT = 2
const HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND = 0

function countNodes(node: { children?: CodeMapNode[] }) {
    let count = 1
    if (node.children) {
        for (const child of node.children) {
            count += countNodes(child)
        }
    }
    return count
}

// TODO this function exists twice in the code - please refactor it.
function calculateSize(node: CodeMapNode, metricName: string) {
    let totalSize = node.attributes[metricName] || 0

    if (totalSize === 0 && node.children && node.children.length > 0) {
        for (const child of node.children) {
            totalSize += calculateSize(child, metricName)
        }
    }
    return totalSize
}

function buildingArrayToMap(highlighted: CodeMapBuilding[]) {
    const geomMap: Map<number, CodeMapBuilding> = new Map()

    for (const building of highlighted) {
        geomMap.set(building.id, building)
    }

    return geomMap
}

function buildRootFolderForFixedFolders(map: CodeMapNode, heightScale: number, state: CcState, isDeltaState: boolean) {
    const flattened = isNodeFlat(map, state)
    const height = FOLDER_HEIGHT
    const width = 100
    const length = 100

    return {
        name: map.name,
        id: 0,
        width,
        height,
        length,
        depth: 0,
        x0: 0,
        z0: 0,
        y0: 0,
        isLeaf: false,
        attributes: map.attributes,
        edgeAttributes: map.edgeAttributes,
        deltas: map.deltas,
        heightDelta: (map.deltas?.[state.dynamicSettings.heightMetric] ?? 0) * heightScale,
        visible: isVisible(map, false, state, flattened),
        path: map.path,
        link: map.link,
        markingColor: getMarkingColor(map, state.fileSettings.markedPackages),
        flat: false,
        color: getBuildingColor(map, state, selectedColorMetricDataSelector(state), isDeltaState, flattened),
        incomingEdgePoint: getIncomingEdgePoint(width, height, length, new Vector3(0, 0, 0), treeMapSize),
        outgoingEdgePoint: getOutgoingEdgePoint(width, height, length, new Vector3(0, 0, 0), treeMapSize)
    } as Node
}

function buildNodeFrom(
    squaredNode: HierarchyRectangularNode<CodeMapNode>,
    heightScale: number,
    maxHeight: number,
    state: CcState,
    isDeltaState: boolean
): Node {
    const mapSizeResolutionScaling = getMapResolutionScaleFactor(state.files)
    const { x0, x1, y0, y1, data } = squaredNode
    const isNodeLeaf = isLeaf(squaredNode)
    const flattened = isNodeFlat(data, state)
    const heightValue = getHeightValue(state, data, maxHeight, flattened)
    const depth = data.path.split("/").length - 2
    const height = isNodeLeaf ? resolveHeightValue(heightValue, heightScale, data, state) * mapSizeResolutionScaling : FOLDER_HEIGHT
    const width = x1 - x0
    const length = y1 - y0
    const z0 = squaredNode.depth * FOLDER_HEIGHT
    const heightDelta = (data.deltas?.[state.dynamicSettings.heightMetric] ?? 0) * heightScale * mapSizeResolutionScaling
    const edgePointHeight = height + (heightDelta < 0 ? Math.abs(heightDelta) : 0)

    return {
        name: data.name,
        id: data.id,
        width,
        height,
        length,
        depth,
        mapNodeDepth: squaredNode.depth,
        x0,
        z0,
        y0,
        isLeaf: isNodeLeaf,
        attributes: data.attributes,
        edgeAttributes: data.edgeAttributes,
        deltas: data.deltas,
        heightDelta,
        visible: isVisible(data, isNodeLeaf, state, flattened),
        path: data.path,
        link: data.link,
        markingColor: getMarkingColor(data, state.fileSettings.markedPackages),
        flat: flattened,
        color: getBuildingColor(data, state, selectedColorMetricDataSelector(state), isDeltaState, flattened),
        incomingEdgePoint: getIncomingEdgePoint(width, edgePointHeight, length, new Vector3(x0, z0, y0), treeMapSize),
        outgoingEdgePoint: getOutgoingEdgePoint(width, edgePointHeight, length, new Vector3(x0, z0, y0), treeMapSize)
    }
}

export function getHeightValue(state: CcState, squaredNode: CodeMapNode, maxHeight: number, flattened: boolean) {
    const mapSizeResolutionScaling = getMapResolutionScaleFactor(state.files)

    if (flattened) {
        return MIN_BUILDING_HEIGHT
    }

    let heightValue = squaredNode.attributes[state.dynamicSettings.heightMetric] || HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND
    heightValue *= mapSizeResolutionScaling

    const heightMetric = state.dynamicSettings.heightMetric
    const attributeDescriptors = state.fileSettings.attributeDescriptors
    const isAttributeDirectionInversed = attributeDescriptors[heightMetric]?.direction === 1

    if (isAttributeDirectionInversed) {
        if (state.appSettings.invertHeight) {
            return heightValue
        }
        return maxHeight - heightValue
    }
    if (state.appSettings.invertHeight) {
        return maxHeight - heightValue
    }
    return heightValue
}

export function resolveHeightValue(heightValue: number, heightScale: number, data: CodeMapNode, state: CcState): number {
    const minimalHeight = data.deltas?.[state.dynamicSettings.heightMetric] ? 0 : MIN_BUILDING_HEIGHT
    return Math.max(Math.abs(heightScale * heightValue), minimalHeight)
}

export function isVisible(squaredNode: CodeMapNode, isNodeLeaf: boolean, state: CcState, flattened: boolean) {
    if (squaredNode.isExcluded || (isNodeLeaf && state.appSettings.hideFlatBuildings && flattened)) {
        return false
    }

    if (state.dynamicSettings.focusedNodePath.length > 0) {
        return squaredNode.path.startsWith(state.dynamicSettings.focusedNodePath[0])
    }

    return true
}

export function getIncomingEdgePoint(width: number, height: number, length: number, vector: Vector3, mapSize: number) {
    if (width > length) {
        return new Vector3(vector.x - mapSize + width / 4, vector.y + height, vector.z - mapSize + length / 2)
    }
    return new Vector3(vector.x - mapSize + width / 2, vector.y + height, vector.z - mapSize + length / 4)
}

export function getOutgoingEdgePoint(width: number, height: number, length: number, vector: Vector3, mapSize: number) {
    if (width > length) {
        return new Vector3(vector.x - mapSize + 0.75 * width, vector.y + height, vector.z - mapSize + length / 2)
    }
    return new Vector3(vector.x - mapSize + width / 2, vector.y + height, vector.z - mapSize + 0.75 * length)
}

export function isNodeFlat(codeMapNode: CodeMapNode, state: CcState) {
    if (codeMapNode.isFlattened) {
        return true
    }

    const searchedNodePaths = searchedNodePathsSelector(state)

    if (searchedNodePaths && state.dynamicSettings.searchPattern?.length > 0) {
        return searchedNodePaths.size === 0 || isNodeNonSearched(codeMapNode, state)
    }

    if (state.appSettings.showOnlyBuildingsWithEdges && state.fileSettings.edges.some(edge => edge.visible)) {
        return nodeHasNoVisibleEdges(codeMapNode, state)
    }

    return false
}

function nodeHasNoVisibleEdges(codeMapNode: CodeMapNode, state: CcState) {
    return (
        codeMapNode.edgeAttributes[state.dynamicSettings.edgeMetric] === undefined ||
        !state.fileSettings.edges.some(edge => codeMapNode.path === edge.fromNodeName || codeMapNode.path === edge.toNodeName)
    )
}

function isNodeNonSearched(squaredNode: CodeMapNode, state: CcState) {
    const searchedNodePaths = searchedNodePathsSelector(state)
    return !searchedNodePaths.has(squaredNode.path)
}

export function getBuildingColor(
    node: CodeMapNode,
    { appSettings, dynamicSettings }: CcState,
    nodeMetricDataRange: MetricMinMax,
    isDeltaState: boolean,
    flattened: boolean
) {
    const { mapColors } = appSettings

    if (isDeltaState) {
        return mapColors.base
    }
    const metricValue = node.attributes[dynamicSettings.colorMetric]

    if (metricValue === undefined) {
        return mapColors.base
    }
    if (flattened) {
        return mapColors.flat
    }

    const { colorRange, colorMode } = dynamicSettings

    if (dynamicSettings["colorMetric"] === "unary") {
        return mapColors.positive
    }

    if (colorMode === ColorMode.absolute) {
        if (metricValue < colorRange.from || colorRange.from === nodeMetricDataRange.maxValue) {
            return mapColors.positive
        }
        if (metricValue < colorRange.to || colorRange.to === nodeMetricDataRange.maxValue) {
            return mapColors.neutral
        }
        return mapColors.negative
    }

    if (colorMode === ColorMode.trueGradient) {
        return gradientCalculator.getColorByTrueGradient(mapColors, colorRange, nodeMetricDataRange, metricValue)
    }

    if (colorMode === ColorMode.focusedGradient) {
        return gradientCalculator.getColorByFocusedGradient(mapColors, colorRange, nodeMetricDataRange, metricValue)
    }

    return gradientCalculator.getColorByWeightedGradient(mapColors, colorRange, nodeMetricDataRange, metricValue)
}

export const TreeMapHelper = {
    countNodes,
    buildingArrayToMap,
    buildRootFolderForFixedFolders,
    calculateSize,
    buildNodeFrom,
    isNodeFlat,
    resolveHeightValue,
    getHeightValue,
    FOLDER_HEIGHT,
    MIN_BUILDING_HEIGHT,
    HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND
}
