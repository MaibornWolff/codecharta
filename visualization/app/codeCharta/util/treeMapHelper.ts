import { getMarkingColor, isLeaf } from "./codeMapHelper"
import { Node, CodeMapNode, State } from "../codeCharta.model"
import { Vector3 } from "three"
import { CodeMapBuilding } from "../ui/codeMap/rendering/codeMapBuilding"
import { HierarchyRectangularNode } from "d3-hierarchy"

const FOLDER_HEIGHT = 2
const MIN_BUILDING_HEIGHT = 2
const HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND = 0

function countNodes(node: { children?: CodeMapNode[] }) {
	let count = 1
	if (node.children) {
		for (const child of node.children) {
			count += countNodes(child)
		}
	}
	return count
}

function buildingArrayToMap(highlighted: CodeMapBuilding[]) {
	const geomMap: Map<number, CodeMapBuilding> = new Map()

	for (const building of highlighted) {
		geomMap.set(building.id, building)
	}

	return geomMap
}

function buildRootFolderForFixedFolders(map: CodeMapNode, heightScale: number, state: State, isDeltaState: boolean) {
	const flattened = isNodeFlat(map, state)
	const height = FOLDER_HEIGHT
	const width = 100
	const length = 100

	return {
		name: map.name,
		id: 0,
		width,
		height,
		length,
		depth: 0,
		x0: 0,
		z0: 0,
		y0: 0,
		isLeaf: false,
		attributes: map.attributes,
		edgeAttributes: map.edgeAttributes,
		deltas: map.deltas,
		heightDelta: (map.deltas?.[state.dynamicSettings.heightMetric] ?? 0) * heightScale,
		visible: isVisible(map, false, state, flattened),
		path: map.path,
		link: map.link,
		markingColor: getMarkingColor(map, state.fileSettings.markedPackages),
		flat: false,
		color: getBuildingColor(map, state, isDeltaState, flattened),
		incomingEdgePoint: getIncomingEdgePoint(width, height, length, new Vector3(0, 0, 0), state.treeMap.mapSize),
		outgoingEdgePoint: getOutgoingEdgePoint(width, height, length, new Vector3(0, 0, 0), state.treeMap.mapSize)
	} as Node
}

function buildNodeFrom(
	squaredNode: HierarchyRectangularNode<CodeMapNode>,
	heightScale: number,
	maxHeight: number,
	s: State,
	isDeltaState: boolean
): Node {
	const isNodeLeaf = !isLeaf(squaredNode)
	const flattened = isNodeFlat(squaredNode.data, s)
	const heightValue = getHeightValue(s, squaredNode, maxHeight, flattened)
	const depth = squaredNode.data.path.split("/").length - 2
	const width = squaredNode.x1 - squaredNode.x0
	const height = Math.abs(isNodeLeaf ? Math.max(heightScale * heightValue, MIN_BUILDING_HEIGHT) : FOLDER_HEIGHT)
	const length = squaredNode.y1 - squaredNode.y0
	const { x0 } = squaredNode
	const { y0 } = squaredNode
	const z0 = depth * FOLDER_HEIGHT

	return {
		name: squaredNode.data.name,
		id: squaredNode.data.id,
		width,
		height,
		length,
		depth,
		x0,
		z0,
		y0,
		isLeaf: isNodeLeaf,
		attributes: squaredNode.data.attributes,
		edgeAttributes: squaredNode.data.edgeAttributes,
		deltas: squaredNode.data.deltas,
		heightDelta: (squaredNode.data.deltas?.[s.dynamicSettings.heightMetric] ?? 0) * heightScale,
		visible: isVisible(squaredNode.data, isNodeLeaf, s, flattened),
		path: squaredNode.data.path,
		link: squaredNode.data.link,
		markingColor: getMarkingColor(squaredNode.data, s.fileSettings.markedPackages),
		flat: flattened,
		color: getBuildingColor(squaredNode.data, s, isDeltaState, flattened),
		incomingEdgePoint: getIncomingEdgePoint(width, height, length, new Vector3(x0, z0, y0), s.treeMap.mapSize),
		outgoingEdgePoint: getOutgoingEdgePoint(width, height, length, new Vector3(x0, z0, y0), s.treeMap.mapSize)
	}
}

function getHeightValue(state: State, squaredNode: HierarchyRectangularNode<CodeMapNode>, maxHeight: number, flattened: boolean) {
	if (flattened) {
		return MIN_BUILDING_HEIGHT
	}

	const heightValue = squaredNode.data.attributes[state.dynamicSettings.heightMetric] || HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND

	if (state.appSettings.invertHeight) {
		return maxHeight - heightValue
	}
	return heightValue
}

function isVisible(squaredNode: CodeMapNode, isNodeLeaf: boolean, state: State, flattened: boolean) {
	if (squaredNode.isExcluded || (isNodeLeaf && state.appSettings.hideFlatBuildings && flattened)) {
		return false
	}

	if (state.dynamicSettings.focusedNodePath.length > 0) {
		return squaredNode.path.startsWith(state.dynamicSettings.focusedNodePath)
	}

	return true
}

function getIncomingEdgePoint(width: number, height: number, length: number, vector: Vector3, mapSize: number) {
	if (width > length) {
		return new Vector3(vector.x - mapSize + width / 4, vector.y + height, vector.z - mapSize + length / 2)
	}
	return new Vector3(vector.x - mapSize + width / 2, vector.y + height, vector.z - mapSize + length / 4)
}

function getOutgoingEdgePoint(width: number, height: number, length: number, vector: Vector3, mapSize: number) {
	if (width > length) {
		return new Vector3(vector.x - mapSize + 0.75 * width, vector.y + height, vector.z - mapSize + length / 2)
	}
	return new Vector3(vector.x - mapSize + width / 2, vector.y + height, vector.z - mapSize + 0.75 * length)
}

function isNodeFlat(codeMapNode: CodeMapNode, state: State) {
	if (codeMapNode.isFlattened) {
		return true
	}

	if (state.dynamicSettings.searchedNodePaths && state.dynamicSettings.searchPattern?.length > 0) {
		return state.dynamicSettings.searchedNodePaths.size === 0 || isNodeNonSearched(codeMapNode, state)
	}

	if (state.appSettings.showOnlyBuildingsWithEdges && state.fileSettings.edges.find(edge => edge.visible)) {
		return nodeHasNoVisibleEdges(codeMapNode, state)
	}

	return false
}

function nodeHasNoVisibleEdges(codeMapNode: CodeMapNode, state: State) {
	return (
		codeMapNode.edgeAttributes[state.dynamicSettings.edgeMetric] === undefined ||
		!state.fileSettings.edges.find(edge => codeMapNode.path === edge.fromNodeName || codeMapNode.path === edge.toNodeName)
	)
}

function isNodeNonSearched(squaredNode: CodeMapNode, state: State) {
	return !state.dynamicSettings.searchedNodePaths.has(squaredNode.path)
}

function getBuildingColor(node: CodeMapNode, state: State, isDeltaState: boolean, flattened: boolean) {
	if (isDeltaState) {
		return state.appSettings.mapColors.base
	}
	const metricValue = node.attributes[state.dynamicSettings.colorMetric]

	if (metricValue === undefined) {
		return state.appSettings.mapColors.base
	}
	if (flattened) {
		return state.appSettings.mapColors.flat
	}
	const mapColorPositive = state.appSettings.whiteColorBuildings
		? state.appSettings.mapColors.lightGrey
		: state.appSettings.mapColors.positive
	if (metricValue < state.dynamicSettings.colorRange.from) {
		return state.appSettings.invertColorRange ? state.appSettings.mapColors.negative : mapColorPositive
	}
	if (metricValue > state.dynamicSettings.colorRange.to) {
		return state.appSettings.invertColorRange ? mapColorPositive : state.appSettings.mapColors.negative
	}
	return state.appSettings.mapColors.neutral
}

export const TreeMapHelper = {
	countNodes,
	buildingArrayToMap,
	buildRootFolderForFixedFolders,
	buildNodeFrom
}
