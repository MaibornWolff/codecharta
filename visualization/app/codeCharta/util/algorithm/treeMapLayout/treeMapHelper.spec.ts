import { TreeMapHelper } from "./treeMapHelper"
import { BlacklistType, CodeMapNode, EdgeVisibility, NodeType, State } from "../../../codeCharta.model"
import { CODE_MAP_BUILDING, STATE } from "../../dataMocks"
import { HierarchyRectangularNode } from "d3-hierarchy"

describe("TreeMapHelper", () => {
	describe("build node", () => {
		let codeMapNode: CodeMapNode
		let squaredNode: HierarchyRectangularNode<CodeMapNode>
		let state: State

		const heightScale = 1
		const maxHeight = 2000
		const isDeltaState = false

		beforeEach(() => {
			codeMapNode = {
				name: "Anode",
				path: "/root/Anode",
				type: NodeType.FILE,
				attributes: { theHeight: 100 },
				isExcluded: false,
				isFlattened: false
			}

			squaredNode = {
				data: codeMapNode,
				value: 42,
				x0: 0,
				y0: 0,
				x1: 400,
				y1: 400
			} as HierarchyRectangularNode<CodeMapNode>

			state = STATE
			state.treeMap.mapSize = 1
			state.dynamicSettings.margin = 15
			state.dynamicSettings.heightMetric = "theHeight"
			state.appSettings.invertHeight = false
			state.dynamicSettings.focusedNodePath = ""
		})

		function buildNode() {
			return TreeMapHelper.buildNodeFrom(squaredNode, heightScale, maxHeight, state, isDeltaState)
		}

		it("minimal", () => {
			expect(buildNode()).toMatchSnapshot()
		})

		it("invertHeight", () => {
			state.appSettings.invertHeight = true
			expect(buildNode()).toMatchSnapshot()
		})

		it("deltas", () => {
			squaredNode.data.deltas = {}
			state.dynamicSettings.heightMetric = "theHeight"
			squaredNode.data.deltas[state.dynamicSettings.heightMetric] = 33
			expect(buildNode()).toMatchSnapshot()
			squaredNode.data.deltas = undefined
		})

		it("given negative deltas the resulting heightDelta also should be negative", () => {
			squaredNode.data.deltas = {}
			squaredNode.data.deltas[state.dynamicSettings.heightMetric] = -33
			expect(buildNode().heightDelta).toBe(-33)
			squaredNode.data.deltas = undefined
		})

		it("should set lowest possible height caused by other visible edge pairs", () => {
			state.fileSettings.edges = [
				{
					fromNodeName: "/root/AnotherNode1",
					toNodeName: "/root/AnotherNode2",
					attributes: {
						pairingRate: 33,
						avgCommits: 12
					},
					visible: EdgeVisibility.both
				}
			]
			expect(buildNode()).toMatchSnapshot()
		})

		it("should set markingColor according to markedPackages", () => {
			const color = "#FF0000"
			state.fileSettings.markedPackages = [
				{
					path: "/root/Anode",
					color
				}
			]
			expect(buildNode().markingColor).toEqual(color)
		})

		it("should set no markingColor according to markedPackages", () => {
			const color = "#FF0000"
			state.fileSettings.markedPackages = [
				{
					path: "/root/AnotherNode",
					color
				}
			]
			expect(buildNode().markingColor).toEqual(undefined)
		})

		it("should be visible if it's a children of the focused node path", () => {
			state.dynamicSettings.focusedNodePath = "/root"
			expect(buildNode().visible).toBeTruthy()
		})

		it("should not be visible if it's excluded", () => {
			codeMapNode.isExcluded = true
			expect(buildNode().visible).toBeFalsy()
		})

		it("should be visible if it's not excluded and no focused node path is given", () => {
			expect(buildNode().visible).toBeTruthy()
		})

		describe("isNodeToBeFlat", () => {
			beforeEach(() => {
				codeMapNode = {
					name: "Anode",
					path: "/root/Anode",
					type: NodeType.FILE,
					attributes: {},
					edgeAttributes: { pairingRate: { incoming: 42, outgoing: 23 } },
					isExcluded: false,
					isFlattened: false
				}

				squaredNode = {
					data: codeMapNode,
					value: 42,
					x0: 0,
					y0: 0,
					x1: 400,
					y1: 400
				} as HierarchyRectangularNode<CodeMapNode>

				state = STATE
				state.treeMap.mapSize = 1
				state.dynamicSettings.margin = 15
			})

			it("should not be a flat node when no visibleEdges", () => {
				state.fileSettings.edges = []
				expect(buildNode().flat).toBeFalsy()
			})

			it("should be a flat node when other edges are visible", () => {
				state.appSettings.showOnlyBuildingsWithEdges = true
				state.fileSettings.edges = [
					{
						fromNodeName: "/root/anotherNode",
						toNodeName: "/root/anotherNode2",
						attributes: {},
						visible: EdgeVisibility.both
					}
				]
				expect(buildNode().flat).toBeTruthy()
			})

			it("should not be a flat node when it contains edges", () => {
				state.fileSettings.edges = [
					{
						fromNodeName: "/root/Anode",
						toNodeName: "/root/anotherNode",
						attributes: {}
					}
				]
				expect(buildNode().flat).toBeFalsy()
			})

			it("should not be a flat node, because its searched for", () => {
				state.dynamicSettings.searchedNodePaths = new Set(["/root/Anode"])
				state.dynamicSettings.searchPattern = "Anode"
				expect(buildNode().flat).toBeFalsy()
			})

			it("should be a flat node, because other nodes are searched for", () => {
				state.dynamicSettings.searchedNodePaths = new Set(["/root/anotherNode", "/root/anotherNode2"])
				state.dynamicSettings.searchPattern = "Anode"
				expect(buildNode().flat).toBeTruthy()
			})

			it("should not be a flat node when searchPattern is empty", () => {
				state.dynamicSettings.searchedNodePaths = new Set(["/root/anotherNode", "/root/anotherNode2"])
				state.dynamicSettings.searchPattern = ""
				expect(buildNode().flat).toBeFalsy()
			})

			it("should be flat if node is flattened in blacklist", () => {
				state.fileSettings.blacklist = [{ path: "*Anode", type: BlacklistType.flatten }]
				squaredNode.data.isFlattened = true

				expect(buildNode().flat).toBeTruthy()
			})

			it("should not be flat if node is not blacklisted", () => {
				state.fileSettings.blacklist = []

				expect(buildNode().flat).toBeFalsy()
			})
		})

		describe("getBuildingColor", () => {
			let node: CodeMapNode

			beforeEach(() => {
				node = {
					name: "Anode",
					path: "/root/Anode",
					type: NodeType.FILE,
					attributes: {}
				} as CodeMapNode

				squaredNode.data = node

				node.attributes = { validMetricName: 0 }

				state = STATE
				state.appSettings.invertColorRange = false
				state.appSettings.whiteColorBuildings = false
				state.dynamicSettings.colorRange.from = 5
				state.dynamicSettings.colorRange.to = 10
				state.dynamicSettings.colorMetric = "validMetricName"
			})

			it("creates grey building for undefined colorMetric", () => {
				state.dynamicSettings.colorMetric = "invalid"
				expect(buildNode().color).toBe(state.appSettings.mapColors.base)
			})

			it("creates flat colored building", () => {
				state.fileSettings.blacklist = [{ path: "*Anode", type: BlacklistType.flatten }]
				squaredNode.data.isFlattened = true

				expect(buildNode().color).toBe(state.appSettings.mapColors.flat)
			})

			it("creates green colored building colorMetricValue < colorRangeFrom", () => {
				expect(buildNode().color).toBe(state.appSettings.mapColors.positive)
			})

			it("creates white colored building colorMetricValue < colorRangeFrom", () => {
				state.appSettings.whiteColorBuildings = true

				expect(buildNode().color).toBe(state.appSettings.mapColors.lightGrey)
			})

			it("creates red colored building colorMetricValue < colorRangeFrom with inverted range", () => {
				state.appSettings.invertColorRange = true

				expect(buildNode().color).toBe(state.appSettings.mapColors.negative)
			})

			it("creates red colored building colorMetricValue > colorRangeFrom", () => {
				node.attributes = { validMetricName: 12 }

				expect(buildNode().color).toBe(state.appSettings.mapColors.negative)
			})

			it("creates green colored building colorMetricValue > colorRangeFrom with inverted range", () => {
				state.appSettings.invertColorRange = true
				node.attributes = { validMetricName: 12 }

				expect(buildNode().color).toBe(state.appSettings.mapColors.positive)
			})

			it("creates white colored building colorMetricValue > colorRangeFrom with inverted range", () => {
				state.appSettings.invertColorRange = true
				state.appSettings.whiteColorBuildings = true
				node.attributes = { validMetricName: 12 }

				expect(buildNode().color).toBe(state.appSettings.mapColors.lightGrey)
			})

			it("creates yellow colored building", () => {
				node.attributes = { validMetricName: 7 }

				expect(buildNode().color).toBe(state.appSettings.mapColors.neutral)
			})
		})
	})

	describe("count nodes", () => {
		it("root only should be 1", () => {
			const root = {}
			expect(TreeMapHelper.countNodes(root)).toBe(1)
		})

		it("root plus child should be 2", () => {
			const root: CodeMapNode = {
				children: [{ name: "Foo", type: NodeType.FILE }],
				name: "root",
				type: NodeType.FOLDER
			}
			expect(TreeMapHelper.countNodes(root)).toBe(2)
		})

		it("root plus child in child should be 3", () => {
			const root: CodeMapNode = {
				children: [{ children: [{ name: "Foo", type: NodeType.FILE }], name: "Foo", type: NodeType.FILE }],
				name: "root",
				type: NodeType.FOLDER
			}
			expect(TreeMapHelper.countNodes(root)).toBe(3)
		})

		it("root plus two children should be 3", () => {
			const root: CodeMapNode = {
				children: [
					{ name: "Foo", type: NodeType.FILE },
					{ name: "Bar", type: NodeType.FILE }
				],
				name: "root",
				type: NodeType.FOLDER
			}
			expect(TreeMapHelper.countNodes(root)).toBe(3)
		})
	})

	describe("buildingArrayToMap", () => {
		it("should convert a array of buildings to a map", () => {
			const result = TreeMapHelper.buildingArrayToMap([CODE_MAP_BUILDING])

			expect(result.get(CODE_MAP_BUILDING.id)).toEqual(CODE_MAP_BUILDING)
		})
	})
})
