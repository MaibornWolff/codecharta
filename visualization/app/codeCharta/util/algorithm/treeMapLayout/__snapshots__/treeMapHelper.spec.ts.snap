// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeMapHelper build node deltas 1`] = `
Object {
  "attributes": Object {
    "theHeight": 100,
  },
  "color": "#666666",
  "deltas": Object {
    "theHeight": 33,
  },
  "depth": 1,
  "edgeAttributes": undefined,
  "flat": false,
  "height": 100,
  "heightDelta": 33,
  "id": undefined,
  "incomingEdgePoint": Vector3 {
    "x": 199,
    "y": 102,
    "z": 99,
  },
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "mapNodeDepth": undefined,
  "markingColor": undefined,
  "name": "Anode",
  "outgoingEdgePoint": Vector3 {
    "x": 199,
    "y": 102,
    "z": 299,
  },
  "path": "/root/Anode",
  "visible": true,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 2,
}
`;

exports[`TreeMapHelper build node invertHeight 1`] = `
Object {
  "attributes": Object {
    "theHeight": 100,
  },
  "color": "#666666",
  "deltas": undefined,
  "depth": 1,
  "edgeAttributes": undefined,
  "flat": false,
  "height": 1900,
  "heightDelta": 0,
  "id": undefined,
  "incomingEdgePoint": Vector3 {
    "x": 199,
    "y": 1902,
    "z": 99,
  },
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "mapNodeDepth": undefined,
  "markingColor": undefined,
  "name": "Anode",
  "outgoingEdgePoint": Vector3 {
    "x": 199,
    "y": 1902,
    "z": 299,
  },
  "path": "/root/Anode",
  "visible": true,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 2,
}
`;

exports[`TreeMapHelper build node minimal 1`] = `
Object {
  "attributes": Object {
    "theHeight": 100,
  },
  "color": "#666666",
  "deltas": undefined,
  "depth": 1,
  "edgeAttributes": undefined,
  "flat": false,
  "height": 100,
  "heightDelta": 0,
  "id": undefined,
  "incomingEdgePoint": Vector3 {
    "x": 199,
    "y": 102,
    "z": 99,
  },
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "mapNodeDepth": undefined,
  "markingColor": undefined,
  "name": "Anode",
  "outgoingEdgePoint": Vector3 {
    "x": 199,
    "y": 102,
    "z": 299,
  },
  "path": "/root/Anode",
  "visible": true,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 2,
}
`;

exports[`TreeMapHelper build node should set lowest possible height caused by other visible edge pairs 1`] = `
Object {
  "attributes": Object {
    "theHeight": 100,
  },
  "color": "#666666",
  "deltas": undefined,
  "depth": 1,
  "edgeAttributes": undefined,
  "flat": false,
  "height": 100,
  "heightDelta": 0,
  "id": undefined,
  "incomingEdgePoint": Vector3 {
    "x": 199,
    "y": 102,
    "z": 99,
  },
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "mapNodeDepth": undefined,
  "markingColor": undefined,
  "name": "Anode",
  "outgoingEdgePoint": Vector3 {
    "x": 199,
    "y": 102,
    "z": 299,
  },
  "path": "/root/Anode",
  "visible": true,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 2,
}
`;
